serenity/qserenityintegration.cpp

```diff
@@ -0,0 +1,167 @@
+    : m_fontDatabase(0)
+{
+    qputenv("QT_ENABLE_REGEXP_JIT", "0");
+    qputenv("QT_QPA_FONTDIR", "/res/fonts");
+
+    if (pledge("stdio recvfd sendfd rpath wpath cpath unix getkeymap fattr prot_exec", nullptr) < 0) {
+        perror("pledge");
+    }
+    else {
+        perror("pledge");
+    }
+    app = GUI::Application::construct(0, nullptr);
+
+    m_primaryScreen = new QSerenityScreen();
+
+    QWindowSystemInterface::handleScreenAdded(m_primaryScreen);
+}
+
+QSerenityIntegration::~QSerenityIntegration()
+{
+    QWindowSystemInterface::handleScreenRemoved(m_primaryScreen);
+    delete m_fontDatabase;
+}
+
+bool QSerenityIntegration::hasCapability(QPlatformIntegration::Capability cap) const
+{
+    switch (cap) {
+    case ThreadedPixmaps: return true;
+    case MultipleWindows: return true;
+    case NonFullScreenWindows: return true;
+    case RhiBasedRendering: return false;
+    case PaintEvents: return true;
+    case WindowManagement: return true;
+    default: return QPlatformIntegration::hasCapability(cap);
+    }
+}
+
+// Dummy font database that does not scan the fonts directory to be
+// used for command line tools like qmlplugindump that do not create windows
+// unless DebugBackingStore is activated.
+class DummyFontDatabase : public QPlatformFontDatabase
+{
+public:
+    virtual void populateFontDatabase() override {}
+};
+
+QPlatformFontDatabase *QSerenityIntegration::fontDatabase() const
+{
+    m_fontDatabase = new QFreeTypeFontDatabase;
+    //m_fontDatabase->addApplicationFont("", "/res/fonts/SerenitySans-Regular.ttf");
+    return m_fontDatabase;
+}
+
+QPlatformWindow *QSerenityIntegration::createPlatformWindow(QWindow *window) const
+{
+    QPlatformWindow *w = new QSerenityWindow(window);
+    w->requestActivateWindow();
+    return w;
+}
+
+QPlatformBackingStore *QSerenityIntegration::createPlatformBackingStore(QWindow *window) const
+{
+    std::cerr << "Creating a new backing store" << std::endl;
+    return new QSerenityBackingStore(window);
+}
+
+QAbstractEventDispatcher *QSerenityIntegration::createEventDispatcher() const
+{
+    return new QSerenityEventDispatcher();
+}
+
+QSerenityIntegration *QSerenityIntegration::instance()
+{
+    return static_cast<QSerenityIntegration *>(QGuiApplicationPrivate::platformIntegration());
+}
+
+QT_END_NAMESPACE
+
+QSerenityEventDispatcher::QSerenityEventDispatcher(QObject *parent)
+    : QUnixEventDispatcherQPA(parent)
+{
+}
+
+QSerenityEventDispatcher::~QSerenityEventDispatcher() {
+
+}
+
+bool QSerenityEventDispatcher::processEvents(QEventLoop::ProcessEventsFlags flags) {
+    processSerenityEvents();
+    // while this whole class is hacking, this is another layer of hacks: clear the WaitForMoreEvents flag to process everything immediatelly
+    // makes everything run way smoother at the cost of eating all the CPU
+    return QUnixEventDispatcherQPA::processEvents(flags & (~QEventLoop::WaitForMoreEvents));
```

Comment:
If you're going towards the polling way as a hack, let's at least put it behind a timer via `Core::Timer`?
That would coalesce things, so maybe store unique flags and run this `processEvents` multiple times when the timer fires?

Reply:
Not really, I just thought it might make your life easier by not maxing out your CPU.

Reply:
For my personal and selfish usage, this is fine (and I don't really expect a ton of people to try this out). The point is to have a working proof of concept in and iterate upon it. This commit is basically just "something that compiles and more or less runs and can be used in the PR in the main Serenity repo". 

I'm planning to do a little more hacking on it during the holidays, to fix the most glaring issues, this being one of them, so I hope it's fine for you as well.

Reply:
Sure, I don't mind one way or another, personally.

Reply:
Cool, thank you. Then I guess this conversation is resolved then.